File: bindings/kotlin/src/main/kotlin/ai/arcships/aimux/ToolCallRepair.kt
/**
 * aimux — host-side tool-call repair (RFC-0035).
 *
 * Core is single-call: an unparseable tool call never fails generation, it
 * comes back as `ToolCall(invalid = true, error = …)`. Repair therefore happens
 * here, after the call returns: [TypedModel] finds the invalid calls, runs the
 * caller's [RepairToolCall] on the JVM, and hands the answer back to three pure
 * native functions that re-validate and patch. Nothing but JSON crosses the C
 * ABI — no callbacks, no handles, no sessions — so a repair function is free to
 * call back into aimux (typically: ask a model to fix the arguments).
 */

package ai.arcships.aimux

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import java.util.concurrent.ExecutionException
import java.util.concurrent.FutureTask

// ─────────────────────────────────────────────────────────────────────────────
// Types
// ─────────────────────────────────────────────────────────────────────────────

/**
 * A provider-shaped tool call, as the model emitted it.
 *
 * Mirrors core `RawToolCall`. Unlike [ToolCall], `input` is the **raw argument
 * text** (the string the provider sent), not a parsed object — that is what a
 * repair function reads and what it returns a replacement for.
 */
@Serializable
data class RawToolCall(
    @SerialName("tool_call_id") val toolCallId: String,
    @SerialName("tool_name") val toolName: String,
    val input: String,
    @SerialName("provider_executed") val providerExecuted: Boolean? = null,
    val dynamic: Boolean? = null,
    @SerialName("thought_signature") val thoughtSignature: String? = null,
    @SerialName("provider_metadata") val providerMetadata: JsonElement? = null,
)

/**
 * Everything a repair function is given about one invalid tool call.
 *
 * Mirrors the AI SDK `repairToolCall` argument. [error] is the serialized
 * `AiMuxError` that made the call invalid (e.g. `{"InvalidToolInput": …}`),
 * carried verbatim from the call rather than re-derived. [inputSchema] is the
 * JSON Schema of the named function tool, or the AI SDK's empty-object schema
 * when the name resolves to no function tool. [messages] and [instructions]
 * are the prompt the call was generated from.
 */
@Serializable
data class ToolCallRepairContext(
    @SerialName("tool_call") val toolCall: RawToolCall,
    val error: JsonElement,
    @SerialName("input_schema") val inputSchema: JsonElement,
    val tools: List<Tool> = emptyList(),
    val messages: List<ModelMessage> = emptyList(),
    val instructions: String? = null,
)

/**
 * A caller-supplied repair attempt for one invalid tool call.
 *
 * Return a replacement [RawToolCall] to have core re-parse and re-validate it,
 * or `null` to leave the call invalid with its original error. Throwing is
 * reported as a failed repair: the call stays invalid and carries a
 * [ToolCallRepairError] whose cause is the exception's message.
 *
 * Called on the thread that issued the generation, outside any native call, so
 * it may block and may itself call aimux.
 */
typealias RepairToolCall = (ToolCallRepairContext) -> RawToolCall?

// ─────────────────────────────────────────────────────────────────────────────
// Raw native entry points — JSON in, JSON out, no handle.
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Build the repair argument for one invalid tool call.
 *
 * @param toolCallJson One `tool_calls` entry with `"invalid": true`.
 * @param promptJson The prompt the call was generated with (the same string
 *   passed to `generate_text` / `stream_text`).
 * @param optsJson The options the call was generated with (same rule).
 * @return A JSON [ToolCallRepairContext], or the literal `null` when the call
 *   was made without a tool set — AI SDK never repairs those, so skip it.
 */
fun toolCallRepairContext(toolCallJson: String, promptJson: String, optsJson: String? = null): String {
    requireJsonRequired("toolCallJson", toolCallJson)
    requireJsonRequired("promptJson", promptJson)
    requireJson("optsJson", optsJson)
    return stringResult("toolCallRepairContext") { out ->
        FFI.lib.aimux_tool_call_repair_context(toolCallJson, promptJson, optsJson, out)
    }
}

/**
 * Resolve one invalid tool call against a repair reply, returning the final
 * `ToolCall` JSON — valid, or invalid carrying a nested `ToolCallRepair` error.
 *
 * @param replyJson `{"type":"repaired","tool_call":…}` | `{"type":"unchanged"}`
 *   | `{"type":"failed","message":…}`.
 * @throws InvalidArgumentError when the options carry no tools, or the call is
 *   not an invalid one.
 */
fun applyToolCallRepair(toolCallJson: String, optsJson: String?, replyJson: String): String {
    requireJsonRequired("toolCallJson", toolCallJson)
    requireJson("optsJson", optsJson)
    requireJsonRequired("replyJson", replyJson)
    return stringResult("applyToolCallRepair") { out ->
        FFI.lib.aimux_apply_tool_call_repair(toolCallJson, optsJson, replyJson, out)
    }
}

/**
 * Apply a repair reply to a whole serialized result (`GenerateTextResult`,
 * `GenerateObjectResult` via its nested `raw`, or `StreamTextResultAggregated`).
 *
 * Rewrites both `tool_calls` and the matching `response_messages` tool-call
 * part — the transcript must agree with the repaired call, or the next turn
 * sends the model the unrepaired arguments again.
 *
 * @throws InvalidArgumentError when `toolCallId` matches no entry, the matched
 *   call is valid, or the options carry no tools.
 */
fun applyToolCallRepairToResult(
    resultJson: String,
    optsJson: String?,
    toolCallId: String,
    replyJson: String,
): String {
    requireJsonRequired("resultJson", resultJson)
    requireJson("optsJson", optsJson)
    requireJsonRequired("replyJson", replyJson)
    return stringResult("applyToolCallRepairToResult") { out ->
        FFI.lib.aimux_apply_tool_call_repair_to_result(resultJson, optsJson, toolCallId, replyJson, out)
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// Host loop — one helper for results, one for stream parts.
// ─────────────────────────────────────────────────────────────────────────────

private const val REPLY_UNCHANGED = """{"type":"unchanged"}"""

/** Run the caller's function and encode its outcome as a `ToolCallRepairReply`. */
private fun repairReply(repair: RepairToolCall, contextJson: String): String {
    val context = try {
        AimuxJson.decodeFromString(ToolCallRepairContext.serializer(), contextJson)
    } catch (e: Exception) {
        throw InvalidArgumentError(
            "failed to decode ToolCallRepairContext: ${e.message ?: e::class.simpleName}",
            cause = e,
        )
    }
    return try {
        val replacement = repair(context) ?: return REPLY_UNCHANGED
        buildJsonObject {
            put("type", "repaired")
            put("tool_call", AimuxJson.encodeToJsonElement(RawToolCall.serializer(), replacement))
        }.toString()
    } catch (e: Exception) {
        // A host exception has no typed counterpart in core; only its message
        // survives, as the cause of ToolCallRepairError.
        buildJsonObject {
            put("type", "failed")
            put("message", e.message ?: e::class.simpleName ?: "tool call repair failed")
        }.toString()
    }
}

private fun isInvalid(call: JsonObject): Boolean =
    (call["invalid"] as? JsonPrimitive)?.booleanOrNull == true

/**
 * Give every invalid tool call in a result document one repair attempt, in
 * order, and return the patched result JSON.
 *
 * Ids are read from the document as it arrived: patching one call never
 * touches another's id, and one attempt per call is the AI SDK rule.
 */
internal fun repairToolCallsInResult(
    resultJson: String,
    promptJson: String,
    optsJson: String?,
    repair: RepairToolCall,
): String {
    val root = AimuxJson.parseToJsonElement(resultJson) as? JsonObject ?: return resultJson
    // GenerateObjectResult nests the text result under `raw`.
    val target = if (root.containsKey("tool_calls")) root else root["raw"] as? JsonObject ?: return resultJson
    val invalid = (target["tool_calls"] as? JsonArray)
        ?.mapNotNull { it as? JsonObject }
        ?.filter(::isInvalid)
        ?: return resultJson

    var patched = resultJson
    for (call in invalid) {
        val contextJson = toolCallRepairContext(call.toString(), promptJson, optsJson)
        if (AimuxJson.parseToJsonElement(contextJson) is JsonNull) continue
        val id = call["tool_call_id"]?.jsonPrimitive?.content ?: continue
        patched = applyToolCallRepairToResult(patched, optsJson, id, repairReply(repair, contextJson))
    }
    return patched
}

/**
 * Wrap [repair] so it runs on a fresh thread while the caller waits.
 *
 * Invariant: the FFI re-entrancy guard is a THREAD-LOCAL flag set on the
 * thread that entered `aimux_stream_text` — the very thread the stream
 * callbacks run on. Any aimux call that blocks on the runtime (every generate
 * / stream entry point) fails there with `AIMUX_E_FFI_REENTRANT_CALL` (204),
 * and asking a model to fix the arguments is the canonical repair. So the
 * hook must not run on the callback thread. The three repair functions
 * themselves are pure and stay where they are. Part ordering is unchanged:
 * the callback thread blocks until the hook returns.
 *
 * [model] is the streaming model, whose read hold is lent to the worker for
 * the duration of the hook: the callback thread is blocked in `task.get()`
 * still holding it, so the handle cannot be dropped, while a fresh read
 * acquisition here would queue behind a waiting `close()` writer (the lock is
 * fair) and deadlock all three. The lend covers [model] only, and only on the
 * worker thread — a thread the hook starts itself still takes the lock.
 */
internal fun offCallbackThread(model: Model, repair: RepairToolCall): RepairToolCall = { context ->
    val task = FutureTask { model.withLentReadHold { repair(context) } }
    Thread(task, "aimux-tool-call-repair").start()
    try {
        task.get()
    } catch (e: ExecutionException) {
        // Rethrow what the hook threw, so "throw means failed" still maps to
        // the hook's own message rather than to the wrapper's.
        throw e.cause ?: e
    }
}

/**
 * Repair an invalid `ToolCall` stream part, returning the replacement part
 * JSON. Any other part (tool-input deltas included) is returned untouched, so
 * deltas still reach the caller immediately — AI SDK behaviour.
 */
internal fun repairToolCallStreamPart(
    partJson: String,
    promptJson: String,
    optsJson: String?,
    repair: RepairToolCall,
): String {
    // StreamPart is externally tagged: {"ToolCall": {…}}.
    val part = AimuxJson.parseToJsonElement(partJson) as? JsonObject ?: return partJson
    val call = part["ToolCall"] as? JsonObject ?: return partJson
    if (!isInvalid(call)) return partJson

    val contextJson = toolCallRepairContext(call.toString(), promptJson, optsJson)
    if (AimuxJson.parseToJsonElement(contextJson) is JsonNull) return partJson
    val repaired = applyToolCallRepair(call.toString(), optsJson, repairReply(repair, contextJson))
    return JsonObject(part + ("ToolCall" to AimuxJson.parseToJsonElement(repaired))).toString()
}

File: bindings/swift/Sources/Aimux/ToolCallRepair.swift
// ToolCallRepair.swift — host-side tool-call repair (RFC-0035).
//
// aimux-core is single-call: an unparseable tool call never fails generation,
// it comes back as a `ToolCall` with `invalid: true` and `error`. Repair is
// therefore post-processing done here, after the generation call returns: the
// binding asks the library for the AI SDK `repairToolCall` argument, hands it
// to the user's closure, and feeds the answer back into a pure native function
// that re-validates and patches. Only JSON crosses the boundary — nothing is
// registered with the C ABI and no native call is in flight while the closure
// runs, so the closure may itself call back into aimux.
//
// OpenAI-shaped output has no `invalid` / `error` to repair from, so
// `generateTextAsOpenAI` repairs the native result and converts it
// (`aimux_generate_text_result_as_openai`); `streamTextAsOpenAI` does not
// reflect repair (RFC-0035 §4).

import CAimuxFFI
import Foundation

// ─────────────────────────────────────────────────────────────────────────────
// Public types
// ─────────────────────────────────────────────────────────────────────────────

/// A tool call as the provider produced it (`aimux_core::RawToolCall`).
///
/// Field-for-field the same as ``ToolCall``, except `input` is the model's raw
/// argument *text* rather than a parsed value — that text is exactly what a
/// repair function needs to see, and what it returns a corrected version of.
public struct RawToolCall: Codable, Equatable {
    public var toolCallId: String
    public var toolName: String
    /// The model's raw argument text, e.g. `#"{"city":"Singapore"}"#`.
    public var input: String
    public var providerExecuted: Bool?
    public var dynamic: Bool?
    public var thoughtSignature: String?
    public var providerMetadata: JSONValue?

    enum CodingKeys: String, CodingKey {
        case toolCallId = "tool_call_id"
        case toolName = "tool_name"
        case input
        case providerExecuted = "provider_executed"
        case dynamic
        case thoughtSignature = "thought_signature"
        case providerMetadata = "provider_metadata"
    }

    public init(toolCallId: String, toolName: String, input: String,
                providerExecuted: Bool? = nil, dynamic: Bool? = nil,
                thoughtSignature: String? = nil, providerMetadata: JSONValue? = nil) {
        self.toolCallId = toolCallId; self.toolName = toolName; self.input = input
        self.providerExecuted = providerExecuted; self.dynamic = dynamic
        self.thoughtSignature = thoughtSignature; self.providerMetadata = providerMetadata
    }
}

/// The argument handed to ``GenerateTextOptions/repairToolCall``. Mirrors the
/// AI SDK `repairToolCall` argument.
public struct ToolCallRepairContext: Codable, Equatable {
    /// The failing call, with the provider's raw argument text.
    public var toolCall: RawToolCall
    /// The failure the call already carries, as externally-tagged wire JSON
    /// (the same encoding as ``ToolCall/error``), e.g.
    /// `{"InvalidToolInput":{"tool_name":…,"tool_input":…,"cause":…}}`.
    public var error: JSONValue
    /// JSON Schema of the named tool. A name that does not resolve to a
    /// function tool yields the AI SDK's default empty-object schema.
    public var inputSchema: JSONValue
    /// The tool set the call was generated with.
    public var tools: [Tool]
    /// The messages the model saw, derived from the same prompt.
    public var messages: [ModelMessage]
    /// The instructions the call was generated with, if any.
    public var instructions: String?

    enum CodingKeys: String, CodingKey {
        case toolCall = "tool_call"
        case error
        case inputSchema = "input_schema"
        case tools, messages, instructions
    }

    public init(toolCall: RawToolCall, error: JSONValue, inputSchema: JSONValue,
                tools: [Tool], messages: [ModelMessage], instructions: String? = nil) {
        self.toolCall = toolCall; self.error = error; self.inputSchema = inputSchema
        self.tools = tools; self.messages = messages; self.instructions = instructions
    }
}

/// A one-shot repair function: return a corrected ``RawToolCall``, `nil` to
/// leave the call invalid as it is, or throw to fail the repair.
public typealias RepairToolCall = (ToolCallRepairContext) throws -> RawToolCall?

/// Carries the repair closure inside ``GenerateTextOptions`` without costing it
/// its synthesized `Codable` / `Equatable`. Two option values never differ by
/// their repair function: a closure has no equality worth comparing, and the
/// function is host-side only — it is never serialized into the options JSON.
struct RepairToolCallBox: Equatable {
    let run: RepairToolCall
    static func == (_: RepairToolCallBox, _: RepairToolCallBox) -> Bool { true }
}

// ─────────────────────────────────────────────────────────────────────────────
// The three native functions (pure, no handle, no I/O)
// ─────────────────────────────────────────────────────────────────────────────

/// Build the repair argument for one invalid tool call.
///
/// - Parameters:
///   - toolCall: One `GenerateTextResult.tool_calls` entry with `"invalid": true`.
///   - prompt: The prompt JSON the call was generated with.
///   - options: The options JSON the call was generated with.
/// - Returns: The serialized ``ToolCallRepairContext``, or the JSON literal
///   `"null"` when the options carried no tool set — such a call is never
///   repaired (AI SDK rule), and that is a success, not a failure.
public func toolCallRepairContext(
    toolCall: String, prompt: String, options: String? = nil
) throws -> String {
    try validateJson(toolCall, parameter: "toolCall")
    try validateJson(prompt, parameter: "prompt")
    try validateJson(options, parameter: "options")
    return try ffiStringCall { aimux_tool_call_repair_context(toolCall, prompt, options, $0) }
}

/// Resolve one invalid tool call against a repair reply, returning the
/// serialized ``ToolCall`` — valid, or invalid carrying a nested
/// `ToolCallRepair` error.
///
/// Throws `AimuxError.invalidArgument` when the options carry no tool set (for
/// such a call ``toolCallRepairContext(toolCall:prompt:options:)`` already
/// answered `null`) or when `toolCall` is not an invalid call.
public func applyToolCallRepair(
    toolCall: String, options: String?, reply: String
) throws -> String {
    try validateJson(toolCall, parameter: "toolCall")
    try validateJson(options, parameter: "options")
    try validateJson(reply, parameter: "reply")
    return try ffiStringCall { aimux_apply_tool_call_repair(toolCall, options, reply, $0) }
}

/// Apply a repair reply to a serialized `GenerateTextResult` or
/// `GenerateObjectResult`. Both `tool_calls` and the matching
/// `response_messages` tool-call part are rewritten — the transcript must keep
/// pointing at the same entry, or the next turn replays the unrepaired
/// arguments.
public func applyToolCallRepairToResult(
    result: String, options: String?, toolCallId: String, reply: String
) throws -> String {
    try validateJson(result, parameter: "result")
    try validateJson(options, parameter: "options")
    try validateJson(reply, parameter: "reply")
    return try ffiStringCall {
        aimux_apply_tool_call_repair_to_result(result, options, toolCallId, reply, $0)
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// The host loop — one helper for results, one for stream parts
// ─────────────────────────────────────────────────────────────────────────────

/// Repair every invalid tool call in a serialized result document and return
/// the patched JSON (the input unchanged when no repair function is set).
///
/// Handles `GenerateTextResult` and `StreamTextResultAggregated` (top-level
/// `tool_calls`) as well as `GenerateObjectResult` (`raw.tool_calls`). Calls
/// are visited in document order and each gets at most one repair attempt, as
/// in the AI SDK.
func repairedResultJson(
    _ resultJson: String, promptJson: String, optsJson: String?,
    options: GenerateTextOptions?
) throws -> String {
    guard let repair = options?.repairToolCall else { return resultJson }
    guard let document = (try? JSONSerialization.jsonObject(with: Data(resultJson.utf8)))
        as? [String: Any] else { return resultJson }
    // A GenerateObjectResult keeps the whole text result under `raw`.
    let target = document["tool_calls"] != nil ? document : document["raw"] as? [String: Any]
    guard let calls = target?["tool_calls"] as? [[String: Any]] else { return resultJson }

    var patched = resultJson
    for call in calls where call["invalid"] as? Bool == true {
        guard let id = call["tool_call_id"] as? String, let callJson = jsonString(call) else { continue }
        let contextJson = try toolCallRepairContext(
            toolCall: callJson, prompt: promptJson, options: optsJson
        )
        guard let context = try decodeRepairContext(contextJson) else { continue }
        patched = try applyToolCallRepairToResult(
            result: patched, options: optsJson, toolCallId: id,
            reply: repairReply(context, repair)
        )
    }
    return patched
}

/// Repair a `{"ToolCall": …}` stream part, returning the replacement part JSON.
///
/// Any other part — `ToolInputDelta` above all — is returned untouched and
/// immediately: the AI SDK forwards argument deltas verbatim even when a repair
/// function is configured.
func repairedStreamPartJson(
    _ partJson: String, promptJson: String, optsJson: String?,
    options: GenerateTextOptions?
) throws -> String {
    guard let repair = options?.repairToolCall,
          let part = (try? JSONSerialization.jsonObject(with: Data(partJson.utf8))) as? [String: Any],
          let call = part["ToolCall"] as? [String: Any],
          call["invalid"] as? Bool == true,
          let callJson = jsonString(call)
    else { return partJson }

    // The two library calls are pure — no runtime, no `ffi_block_on` — so they
    // are safe right here on the callback thread. Only the user's closure has
    // to be moved off it.
    let contextJson = try toolCallRepairContext(
        toolCall: callJson, prompt: promptJson, options: optsJson
    )
    guard let context = try decodeRepairContext(contextJson) else { return partJson }
    let reply = offCallbackThread { repairReply(context, repair) }
    let repairedJson = try applyToolCallRepair(
        toolCall: callJson, options: optsJson, reply: reply
    )
    let repaired = try JSONSerialization.jsonObject(with: Data(repairedJson.utf8))
    return jsonString(["ToolCall": repaired]) ?? partJson
}

/// Run `body` on another thread and block until it answers.
///
/// A dedicated `Thread`, not `DispatchQueue.sync`: GCD runs a `sync` block on
/// the calling thread when it can, which would put the hook right back on the
/// guarded callback thread.
///
/// Invariant: the FFI re-entrancy guard is **thread-local**. A stream
/// `on_part` callback runs on the very thread that entered
/// `aimux_stream_text`, still inside that call's `ffi_block_on`, so any aimux
/// entry point invoked on it comes straight back as
/// `AIMUX_E_FFI_REENTRANT_CALL` (204). A repair function is expected to ask a
/// model to fix the arguments, so it must not run there.
///
/// Ordering is unaffected: the callback thread waits here, so the repaired
/// part is still delivered in its place in the stream, before the next one.
private func offCallbackThread<T>(_ body: @escaping () -> T) -> T {
    var answer: T?
    let done = DispatchSemaphore(value: 0)
    let worker = Thread {
        answer = body()
        done.signal()
    }
    worker.start()
    done.wait()
    // `body` is non-throwing and always assigns before signalling.
    return answer!
}

/// Decode a repair context, or `nil` for the JSON literal `null` the library
/// writes for a call made without a tool set (skip it — never repairable).
private func decodeRepairContext(_ json: String) throws -> ToolCallRepairContext? {
    guard json.trimmingCharacters(in: .whitespacesAndNewlines) != "null" else { return nil }
    return try JSONDecoder().decode(ToolCallRepairContext.self, from: Data(json.utf8))
}

/// Run the user's repair function and phrase its outcome as a reply:
/// a replacement call, `nil` (leave the call as it is), or a throw (the call
/// stays invalid, carrying a `ToolCallRepair` error with this message as cause).
private func repairReply(_ context: ToolCallRepairContext, _ repair: RepairToolCall) -> String {
    do {
        guard let replacement = try repair(context) else { return #"{"type":"unchanged"}"# }
        let encoded = try JSONEncoder().encode(replacement)
        return #"{"type":"repaired","tool_call":"# + String(decoding: encoded, as: UTF8.self) + "}"
    } catch {
        let message = (error as? LocalizedError)?.errorDescription ?? String(describing: error)
        return jsonString(["type": "failed", "message": message])
            ?? #"{"type":"failed","message":"repair function failed"}"#
    }
}

/// Serialize a JSONSerialization-compatible value back to a JSON string.
private func jsonString(_ object: Any) -> String? {
    guard let data = try? JSONSerialization.data(withJSONObject: object, options: [.fragmentsAllowed])
    else { return nil }
    return String(data: data, encoding: .utf8)
}
